src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -779,49 +796,79 @@ private static bool CanConvertMetadataReferenceToProjectReference(Solution solut
-        return reference;
+        return solutionServices.GetRequiredService<IMetadataService>().GetReference(fullFilePath, properties);
     }
 
     private void StartRefreshingMetadataReferencesForFile(object? sender, string fullFilePath)
-    {
-        var asyncToken = WorkspaceListener.BeginAsyncOperation(nameof(StartRefreshingMetadataReferencesForFile));
+        => StartRefreshingReferencesForFile(
+            fullFilePath,
+            static project => project.MetadataReferences.OfType<PortableExecutableReference>(),
+            static reference => reference.FilePath!,
+            static (@this, reference) => CreateMetadataReference_NoLock(reference.FilePath!, reference.Properties, @this.SolutionServices),
+            static (projectUpdateState, oldReference, newReference) => projectUpdateState
+                .WithIncrementalMetadataReferenceRemoved(oldReference)
+                .WithIncrementalMetadataReferenceAdded(newReference),
+            static (solution, projectId, oldReference, newReference) => solution
+                .RemoveMetadataReference(projectId, oldReference)
+                .AddMetadataReference(projectId, newReference));
+
+    private void StartRefreshingAnalyzerReferenceForFile(object? sender, string fullFilePath)
+        => StartRefreshingReferencesForFile(
+            fullFilePath,
+            static project => project.AnalyzerReferences.OfType<AnalyzerFileReference>(),
+            static reference => reference.FullPath,
+            static (@this, reference) => new AnalyzerFileReference(reference.FullPath, @this.SolutionServices.GetRequiredService<IAnalyzerAssemblyLoaderProvider>().GetShadowCopyLoader()),
+            static (projectUpdateState, oldReference, newReference) => projectUpdateState
+                .WithIncrementalAnalyzerReferenceRemoved(oldReference)
+                .WithIncrementalAnalyzerReferenceAdded(newReference),
+            static (solution, projectId, oldReference, newReference) => solution
+                .RemoveAnalyzerReference(projectId, oldReference)
+                .AddAnalyzerReference(projectId, newReference));
 
-        var task = StartRefreshingMetadataReferencesForFileAsync(sender, fullFilePath);
+    /// <summary>
+    /// Core helper that handles refreshing the references we have for a particular <see
+    /// cref="PortableExecutableReference"/> or <see cref="AnalyzerFileReference"/>.
+    /// </summary>
+    private void StartRefreshingReferencesForFile<TReference>(
+        string fullFilePath,
+        Func<Project, IEnumerable<TReference>> getReferences,
+        Func<TReference, string> getFilePath,
+        Func<ProjectSystemProjectFactory, TReference, TReference> createNewReference,
+        Func<ProjectUpdateState, TReference, TReference, ProjectUpdateState> updateState,
+        Func<Solution, ProjectId, TReference, TReference, Solution> updateSolution)
+        where TReference : class
```

Comment:
extracted a common generic helper so we can have the same logic for PERefs and AnalyzerRefs above.